third/nmh/man/forw.man
.\"
.\" %nmhwarning%
.\" $Id: forw.man,v 1.1.1.1 1999-02-07 18:14:19 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH FORW %manext1% MH.6.8 [%nmhversion%]
.SH NAME
forw \- forward messages
.SH SYNOPSIS
.in +.5i
.ti -.5i
forw
\%[+folder] \%[msgs]
.br
\%[\-annotate] \%[\-noannotate]
\%[\-form\ formfile]
.br
\%[\-format] \%[\-noformat]
\%[\-filter\ filterfile]
.br
\%[\-inplace] \%[\-noinplace]
\%[\-mime] \%[\-nomime]
.br
\%[\-draftfolder\ +folder] \%[\-draftmessage\ msg]
.br
\%[\-nodraftfolder]
\%[\-editor\ editor] \%[\-noedit]
.br
\%[\-whatnowproc\ program] \%[\-nowhatnowproc]
.br
\%[\-dashstuffing] \%[\-nodashstuffing]
\%[\-version]
\%[\-help]

.ti .5i
forw
\%[+folder] \%[msgs]
\%[\-digest\ list] \%[\-issue\ number]
.br
\%[\-volume\ number]
\%[other\ switches\ for\ \fIforw\fR]
\%[\-help]
.in -.5i
.SH DESCRIPTION
\fIForw\fR may be used to prepare a message containing other messages.

It constructs the new message from a forms (components) file, with a
body composed of the message(s) to be forwarded.  An editor is invoked
as in \fIcomp\fR, and after editing is complete, the user is prompted
before the message is sent.

The default message form contains the following elements:

.nf
.in +.5i
.ne 10
.eo
.so %etcdir%/forwcomps
.ec
.in -.5i
.fi

If a file named \*(lqforwcomps\*(rq exists in the user's nmh directory,
it will be used instead of this default form.  You may also specify an
alternate forms file with the switch `\-form\ formfile'.

When If the draft already exists, \fIforw\fR will ask you as to the disposition
of the draft.  A reply of \fBquit\fR will abort \fIforw\fR, leaving the
draft intact; \fBreplace\fR will replace the existing draft with a blank
skeleton; and \fBlist\fR will display the draft.

If the `\-annotate' switch is given, each message being forwarded will
be annotated with the lines

     Forwarded:\ date
     Forwarded:\ addrs

where each address list contains as many lines as required.  This
annotation will be done only if the message is sent directly from
\fIforw\fR.  If the message is not sent immediately from \fIforw\fR,
\*(lqcomp\ \-use\*(rq may be used to re\-edit and send the constructed
message, but the annotations won't take place.  Normally annotations
are done inplace in order to preserve any links to the message.  You may
change this by using the '\-noinplace' switch.

See \fIcomp\fR\0(1) for a description of the `\-editor' and `\-noedit'
switches.

Although \fIforw\fR uses a forms (components) file to direct it how to
construct the beginning of the draft, it uses a message filter file to
direct it as to how each forwarded message should be formatted in the
body of the draft.  The filter file for \fIforw\fR should be a standard
form file for \fImhl\fR, as \fIforw\fR will invoke \fImhl\fR to filter
(re\-format) the forwarded messages prior to being output to the body
of the draft.

The switches `\-noformat', `\-format', and `\-filter\ filterfile' specify
which message filter file to use.

If `\-noformat' is specified (this is the default), then each forwarded
message is output into the draft exactly as it appears with no \fImhl\fR
filtering.

If `\-format' is specified, then a default message filter file is used.
This default message filter should be adequate for most users.
This default filter \*(lqmhl.forward\*(rq is:

.nf
.in +.5i
.ne 10
.eo
.so %etcdir%/mhl.forward
.ec
.in -.5i
.fi

If a file named \*(lqmhl.forward\*(rq exists in the user's nmh
directory, it will be used instead of this form.  You may specify an
alternate message filter file with the switch `\-filter\ filterfile'.

Each forwarded message is separated with an encapsulation delimiter.
By default, any dashes in the first column of the forwarded messages
will be prepended with `\-\ ' so that when received, the message is
suitable for bursting by \fIburst\fR\0(1).  This follows the Internet
RFC\-934 guidelines.  You may use the flag `\-nodashstuffing' in order
to suppress this form of quoting to the forwarded messages.

For users of \fIprompter\fR\0(1), by specifying prompter's `-prepend'
switch in the \&.mh\(ruprofile file, any commentary text is entered
before the forwarded messages.  (A major win!)

To use the MIME rules for encapsulation, specify the `\-mime' switch.
This directs \fIforw\fR to generate an \fImhbuild\fR composition file.
Note that nmh will not invoke \fImhbuild\fR automatically, unless you
add this line to your \&.mh\(ruprofile file:
.sp
.in +.5i
automimeproc: 1
.in -.5i
.sp
Otherwise,
you must specifically give the command
.sp
.in +.5i
What now? mime
.in -.5i
.sp
prior to sending the draft.

The `\-draftfolder\ +folder' and `\-draftmessage\ msg' switches invoke
the \fInmh\fR draft folder facility.  This is an advanced (and highly
useful) feature.  Consult the \fImh-draft\fR(5) man page for more
information.

Upon exiting from the editor, \fIforw\fR will invoke the \fIwhatnow\fR
program.  See \fIwhatnow\fR\0(1) for a discussion of available
options.  The invocation of this program can be inhibited by using the
`\-nowhatnowproc' switch.  (In truth of fact, it is the \fIwhatnow\fR
program which starts the initial edit.  Hence, `\-nowhatnowproc' will
prevent any edit from occurring.)

The `\-digest\ list', `\-issue\ number', and `\-volume\ number' switches
implement a digest facility for \fInmh\fR.  Specifying these switches
enables and/or overloads the following escapes:

.sp 1
.nf
.ta \w'Component  'u +\w'Escape  'u +\w'Returns  'u
\fIType\fR	\fIEscape\fR	\fIReturns\fR	\fIDescription\fR
\fIcomponent\fR	\fIdigest\fR	string	Argument to `\-digest'
\fIfunction\fR	\fIcur\fR	integer	Argument to `\-volume'
\fIfunction\fR	\fImsg\fR	integer	Argument to `\-issue'
.re
.fi

Consult the \fBAdvanced Features\fR section of 
the \fInmh\fR User's Manual for more information on making digests.
.Fi
^%etcdir%/forwcomps~^The standard message skeleton
^or <mh\-dir>/forwcomps~^Rather than the standard skeleton
^%etcdir%/digestcomps~^The message skeleton if `\-digest' is given
^or <mh\-dir>/digestcomps~^Rather than the standard skeleton
^%etcdir%/mhl.forward~^The standard message filter
^or <mh\-dir>/mhl.forward~^Rather than the standard filter
^$HOME/\&.mh\(ruprofile~^The user profile
^<mh\-dir>/draft~^The draft file
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Current\-Folder:~^To find the default current folder
.Ps
^Draft\-Folder:~^To find the default draft\-folder
.Ps
^Editor:~^To override the default editor
.Ps
^Msg\-Protect:~^To set mode when creating a new message (draft)
.Ps
^fileproc:~^Program to refile the message
.Ps
^mhlproc:~^Program to filter messages being forwarded
.Ps
^whatnowproc:~^Program to ask the \*(lqWhat now?\*(rq questions
.Sa
\fIProposed Standard for Message Encapsulation\fR (RFC\-934),
.br
mhbuild(1), comp(1), repl(1), send(1), whatnow(1), mh\-format(5)
.De
`+folder' defaults to the current folder
`msgs' defaults to cur
.Ds
`\-noannotate'
.Ds
`\-nodraftfolder'
.Ds
`\-noformat'
.Ds
`\-inplace'
.Ds
`\-dashstuffing'
.Ds
`\-nomime'
.Co
If a folder is given, it will become the current folder.
The first message forwarded will become the current message.
.Bu

If \fIwhatnowproc\fR is \fIwhatnow\fR, then \fIforw\fR uses a built\-in
\fIwhatnow\fR, it does not actually run the \fIwhatnow\fR program.
Hence, if you define your own \fIwhatnowproc\fR, don't call it
\fIwhatnow\fR since \fIforw\fR won't run it.

When \fIforw\fR is told to annotate the messages it forwards, it
doesn't actually annotate them until the draft is successfully sent.
If from the \fIwhatnowproc\fR, you \fIpush\fR instead of \fIsend\fR,
it's possible to confuse \fIforw\fR by re\-ordering the file (e.g.,
by using `folder\0\-pack') before the message is successfully sent.
\fIDist\fR and \fIrepl\fR don't have this problem.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/forw.c

/*
 * forw.c -- forward a message, or group of messages.
 *
 * $Id: forw.c,v 1.1.1.1 1999-02-07 18:14:13 danw Exp $
 */

#include <h/mh.h>
#include <fcntl.h>
#include <h/fmt_scan.h>
#include <zotnet/tws/tws.h>


#define	IFORMAT	"digest-issue-%s"
#define	VFORMAT	"digest-volume-%s"

static struct swit switches[] = {
#define	ANNOSW                 0
    { "annotate", 0 },
#define	NANNOSW                1
    { "noannotate", 0 },
#define	DFOLDSW                2
    { "draftfolder +folder", 0 },
#define	DMSGSW                 3
    { "draftmessage msg", 0 },
#define	NDFLDSW                4
    { "nodraftfolder", 0 },
#define	EDITRSW                5
    { "editor editor", 0 },
#define	NEDITSW                6
    { "noedit", 0 },
#define	FILTSW                 7
    { "filter filterfile", 0 },
#define	FORMSW                 8
    { "form formfile", 0 },
#define	FRMTSW                 9
    { "format", 5 },
#define	NFRMTSW               10
    { "noformat", 7 },
#define	INPLSW                11
    { "inplace", 0 },
#define	NINPLSW               12
    { "noinplace", 0 },
#define MIMESW                13
    { "mime", 0 },
#define NMIMESW               14
    { "nomime", 0 },
#define	DGSTSW                15
    { "digest list", 0 },
#define	ISSUESW               16
    { "issue number", 0 },
#define	VOLUMSW               17
    { "volume number", 0 },
#define	WHATSW                18
    { "whatnowproc program", 0 },
#define	NWHATSW               19
    { "nowhatnowproc", 0 },
#define	BITSTUFFSW            20
    { "dashstuffing", 0 },		/* interface to mhl */
#define	NBITSTUFFSW           21
    { "nodashstuffing", 0 },
#define VERSIONSW             22
    { "version", 0 },
#define	HELPSW                23
    { "help", 4 },
#define	FILESW                24
    { "file file", -4 },		/* interface from msh */

#ifdef MHE
#define	BILDSW                25
    { "build", -5 },			/* interface from mhe */
#endif /* MHE */

    { NULL, 0 }
};

static struct swit aqrnl[] = {
#define	NOSW         0
    { "quit", 0 },
#define	YESW         1
    { "replace", 0 },
#define	LISTDSW      2
    { "list", 0 },
#define	REFILSW      3
    { "refile +folder", 0 },
#define NEWSW        4
    { "new", 0 },
    { NULL, 0 }
};

static struct swit aqrl[] = {
    { "quit", 0 },
    { "replace", 0 },
    { "list", 0 },
    { "refile +folder", 0 },
    { NULL, 0 }
};

static char drft[BUFSIZ];

static char delim3[] =
    "\n------------------------------------------------------------\n\n";
static char delim4[] = "\n------------------------------\n\n";


static struct msgs *mp = NULL;		/* used a lot */


/*
 * static prototypes
 */
static void mhl_draft (int, char *, int, int, char *, char *, int);
static void copy_draft (int, char *, char *, int, int, int);
static void copy_mime_draft (int);
static int build_form (char *, char *, int, int);


int
main (int argc, char **argv)
{
    int msgp = 0, anot = 0, inplace = 1, mime = 0;
    int issue = 0, volume = 0, dashstuff = 0;
    int nedit = 0, nwhat = 0, i, in;
    int out, isdf = 0, msgnum;
    char *cp, *cwd, *maildir, *dfolder = NULL;
    char *dmsg = NULL, *digest = NULL, *ed = NULL;
    char *file = NULL, *filter = NULL, *folder = NULL;
    char *form = NULL, buf[BUFSIZ], value[10];
    char **argp, **arguments, *msgs[MAXARGS];
    struct stat st;

#ifdef	MHE
    int buildsw = 0;
#endif	/* MHE */

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case ANNOSW: 
		    anot++;
		    continue;
		case NANNOSW: 
		    anot = 0;
		    continue;

		case EDITRSW: 
		    if (!(ed = *argp++) || *ed == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nedit = 0;
		    continue;
		case NEDITSW:
		    nedit++;
		    continue;

		case WHATSW: 
		    if (!(whatnowproc = *argp++) || *whatnowproc == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nwhat = 0;
		    continue;
#ifdef MHE
		case BILDSW:
		    buildsw++;	/* fall... */
#endif /* MHE */
		case NWHATSW: 
		    nwhat++;
		    continue;

		case FILESW: 
		    if (file)
			adios (NULL, "only one file at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    file = path (cp, TFILE);
		    continue;
		case FILTSW:
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    filter = getcpy (etcpath (cp));
		    mime = 0;
		    continue;
		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case FRMTSW:
		    filter = getcpy (etcpath (mhlforward));
		    continue;
		case NFRMTSW:
		    filter = NULL;
		    continue;

		case INPLSW: 
		    inplace++;
		    continue;
		case NINPLSW: 
		    inplace = 0;
		    continue;

		case MIMESW:
		    mime++;
		    filter = NULL;
		    continue;
		case NMIMESW: 
		    mime = 0;
		    continue;

		case DGSTSW: 
		    if (!(digest = *argp++) || *digest == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    mime = 0;
		    continue;
		case ISSUESW:
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    if ((issue = atoi (cp)) < 1)
			adios (NULL, "bad argument %s %s", argp[-2], cp);
		    continue;
		case VOLUMSW:
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    if ((volume = atoi (cp)) < 1)
			adios (NULL, "bad argument %s %s", argp[-2], cp);
		    continue;

		case DFOLDSW: 
		    if (dfolder)
			adios (NULL, "only one draft folder at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    dfolder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
				    *cp != '@' ? TFOLDER : TSUBCWF);
		    continue;
		case DMSGSW:
		    if (dmsg)
			adios (NULL, "only one draft message at a time!");
		    if (!(dmsg = *argp++) || *dmsg == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;
		case NDFLDSW: 
		    dfolder = NULL;
		    isdf = NOTOK;
		    continue;

		case BITSTUFFSW: 
		    dashstuff = 1;	/* trinary logic */
		    continue;
		case NBITSTUFFSW: 
		    dashstuff = -1;	/* trinary logic */
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	} else {
	    msgs[msgp++] = cp;
	}
    }

    cwd = getcpy (pwd ());

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (file && (msgp || folder))
	adios (NULL, "can't mix files and folders/msgs");

try_it_again:

#ifdef MHE
    strncpy (drft, buildsw ? m_maildir ("draft")
			  : m_draft (dfolder, NULL, NOUSE, &isdf), sizeof(drft));

    /* Check if a draft already exists */
    if (!buildsw && stat (drft, &st) != NOTOK) {
#else
    strncpy (drft, m_draft (dfolder, dmsg, NOUSE, &isdf), sizeof(drft));

    /* Check if a draft already exists */
    if (stat (drft, &st) != NOTOK) {
#endif	/* MHE */
	printf ("Draft \"%s\" exists (%ld bytes).", drft, (long) st.st_size);
	for (i = LISTDSW; i != YESW;) {
	    if (!(argp = getans ("\nDisposition? ", isdf ? aqrnl : aqrl)))
		done (1);
	    switch (i = smatch (*argp, isdf ? aqrnl : aqrl)) {
		case NOSW: 
		    done (0);
		case NEWSW: 
		    dmsg = NULL;
		    goto try_it_again;
		case YESW: 
		    break;
		case LISTDSW: 
		    showfile (++argp, drft);
		    break;
		case REFILSW: 
		    if (refile (++argp, drft) == 0)
			i = YESW;
		    break;
		default: 
		    advise (NULL, "say what?");
		    break;
	    }
	}
    }

    if (file) {
	/*
	 * Forwarding a file.
         */
	anot = 0;	/* don't want to annotate a file */
    } else {
	/*
	 * Forwarding a message.
	 */
	if (!msgp)
	    msgs[msgp++] = "cur";
	if (!folder)
	    folder = getfolder (1);
	maildir = m_maildir (folder);

	if (chdir (maildir) == NOTOK)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (folder)))
	    adios (NULL, "unable to read folder %s", folder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    adios (NULL, "no messages in %s", folder);

	/* parse all the message ranges/sequences and set SELECTED */
	for (msgnum = 0; msgnum < msgp; msgnum++)
	    if (!m_convert (mp, msgs[msgnum]))
		done (1);
	seq_setprev (mp);	/* set the previous sequence */
    }

    if (filter && access (filter, R_OK) == NOTOK)
	adios (filter, "unable to read");

    /*
     * Open form (component) file.
     */
    if (digest) {
	if (issue == 0) {
	    snprintf (buf, sizeof(buf), IFORMAT, digest);
	    if (volume == 0
		    && (cp = context_find (buf))
		    && ((issue = atoi (cp)) < 0))
		issue = 0;
	    issue++;
	}
	if (volume == 0)
	    snprintf (buf, sizeof(buf), VFORMAT, digest);
	if ((cp = context_find (buf)) == NULL || (volume = atoi (cp)) <= 0)
	    volume = 1;
	if (!form)
	    form = digestcomps;
	in = build_form (form, digest, volume, issue);
    } else {
	if (form) {
	    if ((in = open (etcpath (form), O_RDONLY)) == NOTOK)
		adios (form, "unable to open form file");
	} else {
	    if ((in = open (etcpath (forwcomps), O_RDONLY)) == NOTOK)
		adios (forwcomps, "unable to open default components file");
	    form = forwcomps;
	}
    }

    if ((out = creat (drft, m_gmprot ())) == NOTOK)
	adios (drft, "unable to create");

    /*
     * copy the components into the draft
     */
    cpydata (in, out, form, drft);
    close (in);

    if (file) {
	/* just copy the file into the draft */
	if ((in = open (file, O_RDONLY)) == NOTOK)
	    adios (file, "unable to open");
	cpydata (in, out, file, drft);
	close (in);
	close (out);
    } else {
	/*
	 * If filter file is defined, then format the
	 * messages into the draft using mhlproc.
	 */
	if (filter)
	    mhl_draft (out, digest, volume, issue, drft, filter, dashstuff);
	else if (mime)
	    copy_mime_draft (out);
	else
	    copy_draft (out, digest, drft, volume, issue, dashstuff);
	close (out);

	if (digest) {
	    snprintf (buf, sizeof(buf), IFORMAT, digest);
	    snprintf (value, sizeof(value), "%d", issue);
	    context_replace (buf, getcpy (value));
	    snprintf (buf, sizeof(buf), VFORMAT, digest);
	    snprintf (value, sizeof(value), "%d", volume);
	    context_replace (buf, getcpy (value));
	}

	context_replace (pfolder, folder);	/* update current folder   */
	seq_setcur (mp, mp->lowsel);		/* update current message  */
	seq_save (mp);				/* synchronize sequences   */
	context_save ();			/* save the context file   */
    }

    if (nwhat)
	done (0);
    what_now (ed, nedit, NOUSE, drft, NULL, 0, mp,
	anot ? "Forwarded" : NULL, inplace, cwd);
    done (1);
}


/*
 * Filter the messages you are forwarding, into the
 * draft calling the mhlproc, and reading its output
 * from a pipe.
 */

static void
mhl_draft (int out, char *digest, int volume, int issue,
            char *file, char *filter, int dashstuff)
{
    pid_t child_id;
    int i, msgnum, pd[2];
    char *vec[MAXARGS];
    char buf1[BUFSIZ];
    char buf2[BUFSIZ];
    
    if (pipe (pd) == NOTOK)
	adios ("pipe", "unable to create");

    vec[0] = r1bindex (mhlproc, '/');

    for (i = 0; (child_id = fork()) == NOTOK && i < 5; i++)
	sleep (5);
    switch (child_id) {
	case NOTOK: 
	    adios ("fork", "unable to");

	case OK: 
	    close (pd[0]);
	    dup2 (pd[1], 1);
	    close (pd[1]);

	    i = 1;
	    vec[i++] = "-forwall";
	    vec[i++] = "-form";
	    vec[i++] = filter;

	    if (digest) {
		vec[i++] = "-digest";
		vec[i++] = digest;
		vec[i++] = "-issue";
		snprintf (buf1, sizeof(buf1), "%d", issue);
		vec[i++] = buf1;
		vec[i++] = "-volume";
		snprintf (buf2, sizeof(buf2), "%d", volume);
		vec[i++] = buf2;
	    }

	    /*
	     * Are we dashstuffing (quoting) the lines that begin
	     * with `-'.  We use the mhl default (don't add any flag)
	     * unless the user has specified a specific flag.
	     */
	    if (dashstuff > 0)
		vec[i++] = "-dashstuffing";
	    else if (dashstuff < 0)
		vec[i++] = "-nodashstuffing";

	    if (mp->numsel >= MAXARGS - i)
		adios (NULL, "more than %d messages for %s exec",
			vec[0], MAXARGS - i);

	    /*
	     * Now add the message names to filter.  We can only
	     * handle about 995 messages (because vec is fixed size),
	     * but that should be plenty.
	     */
	    for (msgnum = mp->lowsel; msgnum <= mp->hghsel && i < sizeof(vec) - 1;
			msgnum++)
		if (is_selected (mp, msgnum))
		    vec[i++] = getcpy (m_name (msgnum));
	    vec[i] = NULL;

	    execvp (mhlproc, vec);
	    fprintf (stderr, "unable to exec ");
	    perror (mhlproc);
	    _exit (-1);

	default: 
	    close (pd[1]);
	    cpydata (pd[0], out, vec[0], file);
	    close (pd[0]);
	    pidXwait(child_id, mhlproc);
	    break;
    }
}


/*
 * Copy the messages into the draft.  The messages are
 * not filtered through the mhlproc.  Do dashstuffing if
 * necessary.
 */

static void
copy_draft (int out, char *digest, char *file, int volume, int issue, int dashstuff)
{
    int fd,i, msgcnt, msgnum;
    int len, buflen;
    register char *bp, *msgnam;
    char buffer[BUFSIZ];

    msgcnt = 1;
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	if (is_selected (mp, msgnum)) {
	    if (digest) {
		strncpy (buffer, msgnum == mp->lowsel ? delim3 : delim4,
			sizeof(buffer));
	    } else {
		/* Get buffer ready to go */
		bp = buffer;
		buflen = sizeof(buffer);

		strncpy (bp, "\n-------", buflen);
		len = strlen (bp);
		bp += len;
		buflen -= len;

		if (msgnum == mp->lowsel) {
		    snprintf (bp, buflen, " Forwarded Message%s",
			mp->numsel > 1 ? "s" : "");
		} else {
		    snprintf (bp, buflen, " Message %d", msgcnt);
		}
		len = strlen (bp);
		bp += len;
		buflen -= len;

		strncpy (bp, "\n\n", buflen);
	    }
	    write (out, buffer, strlen (buffer));

	    if ((fd = open (msgnam = m_name (msgnum), O_RDONLY)) == NOTOK) {
		admonish (msgnam, "unable to read message");
		continue;
	    }

	    /*
	     * Copy the message.  Add RFC934 quoting (dashstuffing)
	     * unless given the -nodashstuffing flag.
	     */
	    if (dashstuff >= 0)
		cpydgst (fd, out, msgnam, file);
	    else
		cpydata (fd, out, msgnam, file);

	    close (fd);
	    msgcnt++;
	}
    }

    if (digest) {
	strncpy (buffer, delim4, sizeof(buffer));
    } else {
	snprintf (buffer, sizeof(buffer), "\n------- End of Forwarded Message%s\n\n",
		mp->numsel > 1 ? "s" : "");
    }
    write (out, buffer, strlen (buffer));

    if (digest) {
	snprintf (buffer, sizeof(buffer), "End of %s Digest [Volume %d Issue %d]\n",
		digest, volume, issue);
	i = strlen (buffer);
	for (bp = buffer + i; i > 1; i--)
	    *bp++ = '*';
	*bp++ = '\n';
	*bp = 0;
	write (out, buffer, strlen (buffer));
    }
}


/*
 * Create a mhn composition file for forwarding message.
 */

static void
copy_mime_draft (int out)
{
    int msgnum;
    char buffer[BUFSIZ];

    snprintf (buffer, sizeof(buffer), "#forw [forwarded message%s] +%s",
	mp->numsel == 1 ? "" : "s", mp->foldpath);
    write (out, buffer, strlen (buffer));
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
	if (is_selected (mp, msgnum)) {
	    snprintf (buffer, sizeof(buffer), " %s", m_name (msgnum));
	    write (out, buffer, strlen (buffer));
	}
    write (out, "\n", 1);
}


static int
build_form (char *form, char *digest, int volume, int issue)
{
    int	in;
    int fmtsize;
    register char *nfs;
    char *line, tmpfil[BUFSIZ];
    register FILE *tmp;
    register struct comp *cptr;
    struct format *fmt;
    int dat[5];

    /* Get new format string */
    nfs = new_fs (form, NULL, NULL);
    fmtsize = strlen (nfs) + 256;

    /* Compile format string */
    fmt_compile (nfs, &fmt);

    FINDCOMP (cptr, "digest");
    if (cptr)
	cptr->c_text = digest;
    FINDCOMP (cptr, "date");
    if (cptr)
	cptr->c_text = getcpy(dtimenow (0));

    dat[0] = issue;
    dat[1] = volume;
    dat[2] = 0;
    dat[3] = fmtsize;
    dat[4] = 0;

    strncpy (tmpfil, m_tmpfil (invo_name), sizeof(tmpfil));
    if ((tmp = fopen (tmpfil, "w+")) == NULL)
	adios (tmpfil, "unable to create");
    unlink (tmpfil);
    if ((in = dup (fileno (tmp))) == NOTOK)
	adios ("dup", "unable to");

    if ((line = malloc ((unsigned) fmtsize)) == NULL)
	adios (NULL, "unable to allocate format line storage");
    fmt_scan (fmt, line, fmtsize, dat);
    fputs (line, tmp);
    free (line);
    if (fclose (tmp))
	adios (tmpfil, "error writing");

    lseek (in, (off_t) 0, SEEK_SET);
    return in;
}
